rfd/0237-sub-ca-support.md

```diff
@@ -0,0 +1,608 @@
+      (...)
+      ----- END CERTIFICATE -----
+    # Informative, must match certificate if present.
+    public_key: 8D:53:F4:BF:54:63:B9:B6:C8:C1:84:A4:08:5B:B1:F1:07:67:96:DF
+    # Informative, must match certificate if present.
+    cluster_name: zarquon2
+EOF
+```
+
+### Alice disables the "db-client" override
+
+Disabling an override makes it inactive, falling back to the corresponding
+Teleport self-signed certificate, but retains the configuration. Disables take
+effect immediately.
+
+A disabled override is still considered during CA rotation.
+
+`tctl auth sub-ca disable-override --type=db-client --cluster=mycluster`
+
+### Alice deletes the "db-client" override
+
+Deleting an override removes it completely, making Teleport use its self-signed
+certificate. Deletes take effect immediately.
+
+`tctl auth sub-ca delete-override --type=db-client --cluster=mycluster`
+
+### Alice performs a "db-client" CA rotation
+
+1. Alice starts a rotation and attempts to advance to the `update_clients` step:
+
+    ```shell
+    $ tctl auth rotate --manual --type=db-client --phase=init
+    $ tctl auth rotate --manual --type=db-client --phase=update_clients
+    > ERROR: Found CA overrides for authority "db-client". You must either
+    > supply an override for public key public key "AB:CD:EF:..." or disable
+    > the override.
+    >
+    > tctl auth sub-ca create-csr --public-key='AB:CD:EF:...'
+    > or
+    > tctl auth sub-ca disable-override --public-key='AB:CD:EF:...'
+    ```
+
+1. Alice updates the CA override for "db-client":
+
+    ```shell
+    $ tctl auth sub-ca create-csr --public-key='AB:CD:EF:...'
+    > (CSR PEM)
+
+    # Alice issues certificate from CSR.
+
+    tctl auth sub-ca create-override cert.pem [chain.pem...]
+    ```
+
+1. Alice advances the rotation to the `update_clients` step:
+
+    ```shell
+    $ tctl auth rotate --manual --type=db-client --phase=update_clients`
+    # OK, all overrides are addressed.
+    ```
+
+## Details
+
+The design works similarly to [RFD 0194 - SPIFFE issuer override][rfd0194]: we
+retain the Teleport generated private key but "override" the self-signed
+certificate with an externally-signed one. Retaining the private key and
+overriding only the self-signed certificate simplifies the implementation and
+allows overrides to take effect without a CA rotation.
+
+Unlike RFD 0194, the override is conceptually generic (it may apply to any
+Teleport CA) and works at a more fundamental system layer. For the initial
+release only the "db-client" and "windows" CAs may be targeted, but support
+could be seamlessly expanded in the future. It could even be applied to the
+SPIFFE TLS CA, replacing a workload_identity_x509_issuer_override.
+
+Overrides for client facing CAs, like "db-client" and "windows", take effect
+immediately. The customer holds all necessary certificates before creating the
+override, so they may take the necessary steps (like updating trusted roots).
```

Comment:
I'm looking for some description of what it means for the override to "take effect". If i'm understanding correctly, it means the overridden CA will be used to issue client certs and will be returned instead of the self-signed one for GetCertAuthority response

Reply:
> Do we have to interfere with the response to GetCertAuthority and GetCertAuthorities?

That was my (naive) plan, yes, along with "figure out how to fix the event stream". Maybe I can get away with just changing the RPCs?

Reply:
The rationale I used for the SPIFFE override is that any receiver of CAs (or credentials, even) must be ready to support intermediate certs which are very much inherently new, so specifying a flag or using a different RPC shouldn't be a problem, the original response must be available to older clients anyway.